e2e/tests/api-driven/src/lps/steps.ts

```diff
@@ -0,0 +1,110 @@
+import { strict as assert } from "node:assert";
+import { Given, When, Then, Before, After, World } from "@cucumber/cucumber";
+import { cleanup, getApplications, getLatestMagicLink, login, setup } from "./helpers.js";
+
+export class CustomWorld extends World {
+  sessionIds!: string[];
+  token!: string;
+  email!: string;
+  applications!: { id: string}[];
+}
+
+Before<CustomWorld>("@lps-magic-links", async function () {
+  this.email = "me@example.com";
+
+  const sessionIds = await setup({ email: this.email });
+  this.sessionIds = sessionIds;
+});
+
+After<CustomWorld>("@lps-magic-links", async function () {
+  await cleanup();
+});
+
+Given<CustomWorld>(
+  "a magic link is generated",
+  async function () {
+    await login(this.email);
+    const { token, email, usedAt } = await getLatestMagicLink();
+
+    assert.equal(
+      email,
+      this.email,
+      "Generated magic link does not match test user"
+    );
+    assert(token, "Failed to fetch magic link token");
+    assert.equal(usedAt, null, "Magic link should not be consumed");
+
+    this.token = token;
+  }
+);
+
+When<CustomWorld>(
+  "an invalid token is provided",
+  async function () {
+    const invalidToken = "bca15271-3ebf-4d57-aab2-799bfa3e41ce";
+    this.applications = await getApplications(this.email, invalidToken);
+  }
+);
+
+When<CustomWorld>(
+  "an invalid email address is provided",
+  async function () {
+    this.applications = await getApplications("wrongEmail@example.com", this.token);
+  }
+);
+
+When<CustomWorld>(
+  "the expiry time has passed",
+  async function () {
+    // Sleep over the 3s expiry length in test envs
+    await new Promise((resolve) => setTimeout(resolve, 3_500));
```

Comment:
The tricky bit here is that we can't easily use fake timers as we're within an E2E context and don't have a test framework (we're just using `node:assert`).

I spent a while trying to configure this for a significantly shorter interval (~100ms) but couldn't quite get it to work consistently - we need to test it both expired and unexpired.

I figure a 3.5 second sleep isn't critical within the regression test suite, but clearly this won't be a sustainable pattern if we need to repeat something similar many times.